move cent conversions upstream to source package

fivetran-reneeli opened the pull request:
to be completed

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+# dbt_stripe v0.16.0
+
+## Breaking Changes
+- This package assumes that amount-based fields, which as raw values are represented in the smallest denomination in Stripe, are cent-based. This PR shifts the existing conversion from cents to dollars to further upstream. Previously, currency-related fields were converted in downstream models, but now have been converted directly in staging models. Since currency-related fields now have differen ...
+
+To disable this default conversion, refer to the [README]((https://github.com/fivetran/dbt_stripe/blob/main/README.md#disabling-cent-to-dollar-conversion)) on disabling the `stripe__amount_divide` variable.
```
See notes from the source PR to make similar updates here.

fivetran-joemarkiewicz replied:
Additionally, we should highlight each field in the end models which are seeing this update. This will drastically change the outputs if customers were not previously aware. We should make this obvious to customers.

fivetran-reneeli replied:
Would this change the outputs if the values of the fields remain the same? For example, `total_sales` in customer_overview would still retain the same value, just the division occurs farther upstream. I can make a callout for the  fields where the code has changed though. Which, by the way, you make a good point in that the division didn't occur across all end models originally. Will circle back on that in the other comment.

fivetran-joemarkiewicz commented on README.md:
````diff
@@ -220,6 +220,15 @@ vars:
     stripe__subscription_metadata: ['the', 'list', 'of', 'property', 'fields'] # Note: this is case-SENSITIVE and must match the casing of the property as it appears in the JSON
 ```
 
+#### Disabling Cent to Dollar Conversion
+
+Amount-based fields, such as `amount` and `net`, are typically displayed in the smallest denomination (e.g., cents for USD). By default, these values are automatically converted to dollars by dividing by `100.0`. To disable this conversion and retain the values in their smallest denomination, set the `stripe__amount_divide` variable to `False` as shown below:
+
+```yml
+vars:
+    stripe__amount_divide: False
+```
````
See source PR notes to make the same updates here.

fivetran-joemarkiewicz commented on README.md:
Can you also address Issue #99 in this PR

fivetran-reneeli replied:
address

fivetran-joemarkiewicz commented on packages.yml:
```diff
@@ -1,3 +1,7 @@
-- package: fivetran/stripe_source
-  version: [">=0.12.0", "<0.13.0"]
\ No newline at end of file
+# - package: fivetran/stripe_source
+#   version: [">=0.13.0", "<0.14.0"]
+
+  - git: https://github.com/fivetran/dbt_stripe_source.git
+    revision: feature/standardize_cent_conversions
+    warn-unpinned: false
```
Reminder to swap before merge

fivetran-joemarkiewicz commented on models/stripe__customer_overview.sql:
```diff
@@ -148,18 +148,18 @@ with balance_transaction_joined as (
-      coalesce(transactions_grouped.total_gross_transaction_amount/100.0, 0) as total_gross_transaction_amount,
-      coalesce(transactions_grouped.total_fees/100.0, 0) as total_fees,
-      coalesce(transactions_grouped.total_net_transaction_amount/100.0, 0) as total_net_transaction_amount,
+      coalesce(transactions_grouped.total_sales, 0) as total_sales,
```
Why did we only divide by 100 for the `int_stripe__account_daily` and `stripe__customer_overview` models? We didn't do this function in the other (arguably more popular) end models such as balance transaction. Are we missing something here? Should we consider retaining the smallest unit by default but instead allow the customer to enable it if they want?

fivetran-reneeli replied:
Oh hmm. I did some digging back into the commit histories of the models. I think this is product of when each model was introduced. For instance, the [int_stripe__account_daily](https://github.com/fivetran/dbt_stripe/commit/1f3c7aa59d168ca46d3d36e19ebc85ac5e2323e1) was created with the /100 in place. But some of the older models never had it so my guess is there wasn't a retroactive update applied to those. 

Are you saying to set the conversion to False by default? I guess that would make more sense from an impact perspective if more fields weren't ever converted than there were. That way, most of the models wouldn't be impacted than if the conversion was set to True.

If that is the case, this means we keep the changes made to `int_stripe__account_daily` and `stripe__customer_overview`, but I will update the documentation and run script accordingly.

fivetran-joemarkiewicz replied:
Thanks for the investigation here! This is really interesting and a bit confusing why we would apply this divide by 100 approach for a select few models, but not others. 

Given this I would request we take the following approach. We have the default state of the `stripe__convert_values` variable be `false`. We then apply an addition conditional to these end models with the divide by 100 to only divide by 100 if the variable is false. This way we can ensure this update will be backwards compatible and not cause major breaking changes. So these end models would look like the following.

```sql
{{ 'coalesce(transactions_grouped.total_refunds/100.0, 0)' if not var('stripe__convert_values', false) else coalesce(transactions_grouped.total_refunds, 0)' }} as total_refunds,
```

fivetran-reneeli replied:
Updated and added a note to the CHANGELOG.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli thanks for working through this PR. I have a few comments and requests. In particular, I also would like to discuss some more why we only divided by 100 in a few particular places in previous versions of this package. I want to make sure we aren't missing anything before we apply an update like this.

fivetran-joemarkiewicz commented on README.md:
````diff
@@ -220,6 +221,15 @@ vars:
+
+```yml
+vars:
+    stripe__convert_values: False
````
Based on my other comment, let's change this default to false and ensure the docs show how to change this to true by default.

fivetran-reneeli replied:
updated

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,16 @@
+# dbt_stripe v0.16.0
```
We'll need to update this based on the new variable being set to false now. We can likely remove the breaking change impacts but mention that the value of this variable will change in the future and they should be made aware, but if they want to set the variable to true then they should know the impact.

fivetran-reneeli replied:
downgraded the version to 0.15.2 and made changes to the wording

fivetran-reneeli replied:
returning to v0.16.0 and updated the changelog notes

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli thanks for these edits, a few more comments in the review. Thanks!

fivetran-reneeli commented:
@fivetran-joemarkiewicz  ready for re-review! I made the /100 disabled by default, with the exception of where I added the backward compatibility. I also added a consistency test to customer_overview and switched the run script to test for `stripe__convert_values: true`.  And updated the relevant documentation.

fivetran-reneeli commented on models/stripe__customer_overview.sql:
Discussed during a team huddle recently and team brought up a good point about how this may be misleading if a customer did intend to turn off the `convert_values` but instead still this model includes the division because of the inverse logic. So this may not be the best approach. I can workshop another strategy

fivetran-joemarkiewicz reviewed (changes requested):
Generally this is looking really good! I have a few comments below before approving

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,6 +1,27 @@
-# dbt_stripe version.version
+# dbt_stripe v0.15.2
```
updated

fivetran-reneeli commented on README.md:
```diff
@@ -74,7 +75,7 @@ Include the following stripe package version in your `packages.yml` file:
 packages:
   - package: fivetran/stripe
-    version: [">=0.15.0", "<0.16.0"]
+    version: [">=0.16.0", "<0.17.0"]
```
updated

fivetran-reneeli replied:
now back to [">=0.16.0", "<0.17.0"] because of the shift to breaking change

fivetran-reneeli commented on integration_tests/tests/consistency/consistency_customer_overview.sql:
```diff
@@ -0,0 +1,70 @@
+        select
+            {% for col in cols %}
+                {% if not loop.first %}, {% endif %}
+                floor(sum({{ col }})) as summed_{{ col }} -- floor and sum is to keep consistency between dev and prod aggs
```
Good idea to test with that schema-- it came down to incorrectly placed logic in the customer_overview model, I fixed it and the validation tests now all pass with that schema.

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,6 +1,26 @@
+# dbt_stripe v0.15.2
+
+## Feature Update: Optionally convert Amounts to Major Units
+- Stripe passes amount-based fields, such as `amount`, `net`, and `fee`, in the smallest denomination as raw form. This means, if your currency has minor and major units such as USD, 100 represents 100 cents, the minor unit, or 1 USD, the major unit. Alternatively, if your currency doesn't use minor units such as JPY, 100 represents 100 JPY. 
```
```suggestion
## Breaking Change
- The aggregated amount-based fields in `stripe__daily_overview` and `stripe__customer_overview` now reflect the raw smallest units (e.g., cents), following Stripe's raw data, instead of converted dollar amounts. This change standardizes values across all models.  

  **Impact**: Customers using these models should note that the values will appear inflated compared to the previous dollar-based representation.  

- Important to note:  
  - The cent-to-dollar conversion has been moved upstream and is now controlled by a new variable: `stripe__convert_values` which is disabled by default.  
  - To enable the conversion and divide all amount-based fields by 100, set `stripe__convert_values` to `True` in your `project.yml`.  
  - See below for more context behind these changes and for detailed setup instructions, see the [README](https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion).  

## Feature Update: Optional Conversion to Major Units
- **Background**:  Stripe provides amount-based fields (e.g., `amount`, `net`, `fee`) in the smallest denomination. For currencies with minor and major units (e.g., USD), 100 represents 100 cents (1 USD). For currencies without minor units (e.g., JPY), 100 represents 100 JPY.  
```

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,6 +1,26 @@
+
+- This PR introduces a variable `stripe__convert_values` (disabled by default) upstream in the staging models which allows users the option to divide all amount-based fields by 100.
+
+- If your data is using a currency with major and minor units, it may be useful to enable this variable if you wish to divide amounts by 100 and therefore convert values into major units. For information on how to do so, refer to the [README]((https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion)) on enabling the `stripe__convert_values` variable.
```
```suggestion
- If your data is using a currency with major and minor units, it may be useful to enable this variable if you wish to divide amounts by 100 and therefore convert values into major units. For information on how to do so, refer to the [README]((https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion)) on enabling the `stripe__convert_values` variable.
  - Examples of currencies using minor units (in which enabling `stripe__convert_values` may be useful) include USD (100 = 1 USD), EUR (100 = 1 Euro), and CAD (100 = 1 CAD).  
```

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,6 +1,26 @@
+
+- If your data is using a currency with major and minor units, it may be useful to enable this variable if you wish to divide amounts by 100 and therefore convert values into major units. For information on how to do so, refer to the [README]((https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion)) on enabling the `stripe__convert_values` variable.
+
+- If your Stripe data is *not* using a currency involving minor units, it may make more sense to retain the amount-based fields as raw form and you may run the package without additional configuration.
```
```suggestion
- If your Stripe data is *not* using a currency involving minor units, it may make more sense to retain the amount-based fields as raw smallest units by leaving `stripe__convert_values` disabled.
  - Examples of currencies NOT using minor units include JPY (100 = 100 JPY), IDR (100 = 100 IDR), KRW (100 = 100 KRW).
```

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,6 +1,26 @@
+
+## Under the Hood
+- Added logic to maintain backwards compatibility in `int_stripe__account_daily` and `stripe__customer_overview` by accounting for the previously existing division by 100. 
+- Updated the `run_models.sh` script to test for when `stripe__convert_values` is set to True.
```
```suggestion

```

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,6 +1,26 @@
+- Updated the descriptions for all amount-based fields to specify the grain of the values and add information about the `stripe__convert_values` variable.
 - Added Quickstart model counts to README. ([#103](https://github.com/fivetran/dbt_stripe/pull/103))
 - Corrected references to connectors and connections in the README. ([#103](https://github.com/fivetran/dbt_stripe/pull/103))
 
```
```suggestion

## Under the Hood
- Updated the `run_models.sh` script to test for when `stripe__convert_values` is set to True.
- Added a consistency test for `stripe__customer_overview`.

```

fivetran-reneeli reviewed (commented):
I accidentally stored my changes as suggestions, so I need to submit my own review in order to commit them

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,9 +1,41 @@
-# dbt_stripe version.version
+# dbt_stripe v0.16.0
+## Breaking Change
+- The aggregated amount-based fields in `stripe__daily_overview` and `stripe__customer_overview` now reflect the raw smallest units (e.g., cents), following Stripe's raw data, instead of converted dollar amounts. This change standardizes values across all models.  
+
+  **Impact**: Customers using these models should note that the values will appear inflated compared to the previous dollar-based representation.  
+
+- Important to note:  
+  - The cent-to-dollar conversion has been moved upstream and is now controlled by a new variable: `stripe__convert_values` which is disabled by default.  
+  - To enable the conversion and divide all amount-based fields by 100, set `stripe__convert_values` to `True` in your `project.yml`.  
+  - See below for more context behind these changes and for detailed setup instructions, see the [README](https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion).  
+
+## Feature Update: Optional Conversion to Major Units
+- **Background**:  Stripe provides amount-based fields (e.g., `amount`, `net`, `fee`) in the smallest denomination. For currencies with minor and major units (e.g., USD), 100 represents 100 cents (1 USD). For currencies without minor units (e.g., JPY), 100 represents 100 JPY.  
+
+- This PR introduces a variable `stripe__convert_values` (disabled by default) upstream in the staging models which allows users the option to divide all amount-based fields by 100.
+
+- If your data is using a currency with major and minor units, it may be useful to enable this variable if you wish to divide amounts by 100 and therefore convert values into major units. For information on how to do so, refer to the [README]((https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion)) on enabling the `stripe__convert_values` variable.
+  - Examples of currencies using minor units (in which enabling `stripe__convert_values` may be useful) include USD (100 = 1 USD), EUR (100 = 1 Euro), and CAD (100 = 1 CAD).  
+
+- If your Stripe data is *not* using a currency involving minor units, it may make more sense to retain the amount-based fields as raw smallest units by leaving `stripe__convert_values` disabled.
+  - Examples of currencies NOT using minor units include JPY (100 = 100 JPY), IDR (100 = 100 IDR), KRW (100 = 100 KRW).
+
+## Notes
+- Currently this package does not support multiple currencies, but we have created a [feature request to support multiple currencies](https://github.com/fivetran/dbt_stripe/issues/102) where you are welcome to provide feedback or contribute to the discussion.
```
Suggestions to consolidate the CHANGELOG since it was becoming a bit repetitive and lengthy while also making sure it's messaging was consistent with what is listed in the source PR.
```suggestion
## Breaking Change
- The aggregated net, gross, sale, charge, refund, and fee amount-based fields in `stripe__daily_overview` and `stripe__customer_overview` now reflect the raw smallest units (e.g., cents), following Stripe's raw data, instead of converted dollar amounts. This change standardizes values across all models.  
  - Customers using these models should note that the values will appear inflated compared to the previous dollar-based representation.  
  - The cent-to-dollar conversion has been moved upstream and is now controlled by a new variable: `stripe__convert_values` which is disabled by default.  
  - See the below Feature Update notes for more context behind these changes and for detailed setup instructions, see the [README](https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion).  

## Feature Update: Optional Conversion to Major Units
Stripe passes amount-based fields, such as `amount`, `net`, and `fee`, in the smallest denomination as raw form. This means, if your currency has minor and major units such as USD, 100 represents 100 cents, the minor unit, or 1 USD, the major unit. Alternatively, if your currency doesn't use minor units such as JPY, 100 represents 100 JPY. 

- This PR introduces a variable `stripe__convert_values` (disabled by default) upstream in the staging models which allows users the option to divide all amount-based fields by 100.
  - For information on how to enable the division, refer to the [README](https://github.com/fivetran/dbt_stripe?tab=readme-ov-file#enabling-cent-to-dollar-conversion) on configuring the `stripe__convert_values` variable.
  - Otherwise, amount-based fields will be brought through in their raw form.
  - Examples of currencies using minor units (in which enabling `stripe__convert_values` is relevant) include United States Dollar (USD), Euro (EUR), and the Canadian Dollar (CAD).
  - Examples of currencies NOT using minor units (in which it makes more sense to keep the amount-based fields in raw form) include Japanese Yen (JPY), Indonesian Rupiah (IDR), and Korean Won (KRW).

## Notes
- This update is not providing multiple currency support; however, we're interested in exploring this functionality if there's a need. As such, we have created a [feature request to support multiple currencies](https://github.com/fivetran/dbt_stripe/issues/102) where you are welcome to provide feedback or contribute to the discussion.
```

fivetran-joemarkiewicz commented on models/stripe__customer_overview.sql:
```diff
@@ -148,18 +148,18 @@ with balance_transaction_joined as (
-      coalesce(transactions_grouped.total_gross_transaction_amount/100.0, 0) as total_gross_transaction_amount,
-      coalesce(transactions_grouped.total_fees/100.0, 0) as total_fees,
-      coalesce(transactions_grouped.total_net_transaction_amount/100.0, 0) as total_net_transaction_amount,
+      {{ 'coalesce(transactions_grouped.total_sales, 0)' if not var('stripe__convert_values', false) else 'coalesce(transactions_grouped.total_sales, 0)' }} as total_sales,
```
It doesn't look like these have been updated. We should remove this logic and ensure the `/100` is no longer present for the impacted fields in this model for this release.

fivetran-reneeli replied:
My bad, I was focused on removing the /100 and forgot to ensure the remainder of the conditional block was removed as well.

fivetran-joemarkiewicz commented on packages.yml:
```diff
@@ -1,3 +1,7 @@
-  version: [">=0.12.0", "<0.13.0"]
\ No newline at end of file
+# - package: fivetran/stripe_source
+#   version: [">=0.12.0", "<0.13.0"]
```
Reminder to update since the source will now be a breaking change.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli thanks for these changes and all the back and forth. A few more change requests and suggestions before approval

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,9 +1,35 @@
 - Added Quickstart model counts to README. ([#103](https://github.com/fivetran/dbt_stripe/pull/103))
 - Corrected references to connectors and connections in the README. ([#103](https://github.com/fivetran/dbt_stripe/pull/103))
 
+## Under the Hood
+- Updated the `run_models.sh` script to test for when `stripe__convert_values` is set to True.
+- Added a consistency test for `stripe__customer_overview`.
+
```
Looks like a duplicate section that can be removed
```suggestion
```

fivetran-joemarkiewicz reviewed (approved):
One small changelog update request and reminder to bump docs before merging.

fivetran-jamie reviewed (approved):
left a suggestion or two in the changelog but otherwise this is approved
